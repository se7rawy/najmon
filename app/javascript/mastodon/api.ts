import type { AxiosResponse, Method, RawAxiosRequestHeaders } from 'axios';
import axios from 'axios';
import LinkHeader from 'http-link-header';

import { getAccessToken } from './initial_state';
import ready from './ready';

export const getLinks = (response: AxiosResponse) => {
  const value = response.headers.link as string | undefined;

  if (!value) {
    return new LinkHeader();
  }

  return LinkHeader.parse(value);
};

const csrfHeader: RawAxiosRequestHeaders = {};

const setCSRFHeader = () => {
  const csrfToken = document.querySelector<HTMLMetaElement>(
    'meta[name=csrf-token]',
  );

  if (csrfToken) {
    csrfHeader['X-CSRF-Token'] = csrfToken.content;
  }
};

void ready(setCSRFHeader);

const authorizationTokenFromInitialState = (): RawAxiosRequestHeaders => {
  const accessToken = getAccessToken();

  if (!accessToken) return {};

  return {
    Authorization: `Bearer ${accessToken}`,
  };
};

// eslint-disable-next-line import/no-default-export
export default function api(withAuthorization = true) {
  return axios.create({
    headers: {
      ...csrfHeader,
      ...(withAuthorization ? authorizationTokenFromInitialState() : {}),
    },

    transformResponse: [
      function (data: unknown) {
        try {
          return JSON.parse(data as string) as unknown;
        } catch {
          return data;
        }
      },
    ],
  });
}

export async function apiRequest<ApiResponse = unknown>(
  method: Method,
  url: string,
  params?: Record<string, unknown>,
) {
  const { data } = await api().request<ApiResponse>({
    method,
    url: '/api/' + url,
    data: params,
  });

  return data;
}
