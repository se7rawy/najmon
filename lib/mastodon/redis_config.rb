# frozen_string_literal: true

def setup_redis_env_url(prefix = nil, defaults = true)
  prefix = "#{prefix.to_s.upcase}_" unless prefix.nil?
  prefix = '' if prefix.nil?

  return if ENV["#{prefix}REDIS_URL"].present?

  password = ENV.fetch("#{prefix}REDIS_PASSWORD") { '' if defaults }
  host     = ENV.fetch("#{prefix}REDIS_HOST") { 'localhost' if defaults }
  port     = ENV.fetch("#{prefix}REDIS_PORT") { 6379 if defaults }
  db       = ENV.fetch("#{prefix}REDIS_DB") { 0 if defaults }

  ENV["#{prefix}REDIS_URL"] = begin
    if [password, host, port, db].all?(&:nil?)
      ENV['REDIS_URL']
    else
      Addressable::URI.parse("redis://#{host}:#{port}/#{db}").tap do |uri|
        uri.password = password if password.present?
      end.normalize.to_str
    end
  end

  return unless ENV["#{prefix}REDIS_SENTINEL"] || ENV['REDIS_SENTINEL']

  ENV["#{prefix}REDIS_SENTINEL"] = ENV["#{prefix}REDIS_SENTINEL"] || ENV['REDIS_SENTINEL']

  sentinel_master = ENV.fetch("#{prefix}REDIS_SENTINEL_MASTER", 'mymaster')
  password = ENV["#{prefix}REDIS_PASSWORD"] || ENV.fetch('REDIS_PASSWORD')

  unless ENV["#{prefix}REDIS_SENTINEL"].include? ','
    ips = Resolv.getaddresses(ENV["#{prefix}REDIS_SENTINEL"])
    port = ENV.fetch("#{prefix}REDIS_SENTINEL_PORT", '26379')

    ENV["#{prefix}REDIS_SENTINEL"] = ips.map do |ip|
      "#{ip}:#{port}"
    end.join(',').concat(',')
  end

  ENV["#{prefix}REDIS_URL"] = "redis://:#{password}@#{sentinel_master}"
end

setup_redis_env_url
setup_redis_env_url(:cache, false)
setup_redis_env_url(:sidekiq, false)

namespace         = ENV.fetch('REDIS_NAMESPACE', nil)
cache_namespace   = namespace ? "#{namespace}_cache" : 'cache'
sidekiq_namespace = namespace

REDIS_CACHE_PARAMS = {
  driver: :hiredis,
  url: ENV['CACHE_REDIS_URL'],
  expires_in: 10.minutes,
  namespace: "#{cache_namespace}:7.1",
  connect_timeout: 5,
  pool: {
    size: Sidekiq.server? ? Sidekiq[:concurrency] : Integer(ENV['MAX_THREADS'] || 5),
    timeout: 5,
  },

  master_name: (ENV.fetch('CACHE_REDIS_SENTINEL_MASTER', 'mymaster') if ENV['CACHE_REDIS_SENTINEL']),
  sentinels: (if ENV['CACHE_REDIS_SENTINEL']
                ENV['CACHE_REDIS_SENTINEL'].split(',').map do |server|
                  host, port = server.split(':')
                  { host: host, port: port.to_i }
                end
              end),
}.freeze

REDIS_SIDEKIQ_PARAMS = {
  driver: :hiredis,
  url: ENV['SIDEKIQ_REDIS_URL'],
  namespace: sidekiq_namespace,

  master_name: (ENV.fetch('SIDEKIQ_REDIS_SENTINEL_MASTER', 'mymaster') if ENV['SIDEKIQ_REDIS_SENTINEL']),
  sentinels: (if ENV['SIDEKIQ_REDIS_SENTINEL']
                ENV['SIDEKIQ_REDIS_SENTINEL'].split(',').map do |server|
                  host, port = server.split(':')
                  { host: host, port: port.to_i }
                end
              end),
}.freeze

ENV['REDIS_NAMESPACE'] = "mastodon_test#{ENV['TEST_ENV_NUMBER']}" if Rails.env.test?
