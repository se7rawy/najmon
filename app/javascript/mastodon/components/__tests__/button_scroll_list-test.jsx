import renderer from 'react-test-renderer';

import { render, screen } from 'mastodon/test_helpers';

import ButtonScrollList from '../button_scroll_list';

jest.mock('mastodon/components/icon', () => {
  return {
    Icon: () => <div>MockIcon</div>,
  };
});

describe('<ButtonScrollList />', () => {
  it('renders an empty button scroll list element', () => {
    const children = [];
    const component = renderer.create(
      <ButtonScrollList>{children}</ButtonScrollList>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders the children', () => {
    const children = Array.from({ length: 5 }, (_, i) => (
      <div key={i} ref={jest.fn()} />
    ));
    const component = renderer.create(
      <ButtonScrollList>{children}</ButtonScrollList>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('scrolls left', () => {
    const children = Array.from({ length: 5 }, (_, i) => (
      <div key={i} ref={jest.fn()} />
    ));
    const component = renderer.create(
      <ButtonScrollList>{children}</ButtonScrollList>,
    );
    const instance = component.getInstance();
    instance.scrollLeft();
  });

  it('scrolls right', () => {
    const children = Array.from({ length: 5 }, (_, i) => (
      <div key={i} ref={jest.fn()} />
    ));
    const component = renderer.create(
      <ButtonScrollList>{children}</ButtonScrollList>,
    );
    const instance = component.getInstance();
    instance.scrollRight();
  });

  it('scrolls left and right correctly', () => {
    const children = Array.from({ length: 10 }, (_, i) => (
      <div key={i}>{i}</div>
    ));
    const component = renderer.create(
      <ButtonScrollList>{children}</ButtonScrollList>,
    );

    setTimeout(() => {
      let instance = component.getInstance();
      instance.scrollRight();
      expect(instance.slide).toBe(1);
    }, 1000);

    setTimeout(() => {
      let instance = component.getInstance();
      instance.scrollLeft();
      expect(instance.slide).toBe(0);
    }, 2000);
  });

  it('handles a single child correctly', () => {
    const children = [<div key={0} ref={jest.fn()} />];
    const component = renderer.create(
      <ButtonScrollList>{children}</ButtonScrollList>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('handles a large number of children correctly', () => {
    const children = Array.from({ length: 50 }, (_, i) => (
      <div key={i} ref={jest.fn()} />
    ));
    const component = renderer.create(
      <ButtonScrollList>{children}</ButtonScrollList>,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('checks if scroll buttons are accessible', () => {
    const children = Array.from({ length: 5 }, (_, i) => (
      <div key={i} ref={jest.fn()} />
    ));
    render(<ButtonScrollList>{children}</ButtonScrollList>);

    const leftButton = screen.getByRole('button', { name: /scroll left/i });
    const rightButton = screen.getByRole('button', { name: /scroll right/i });

    expect(leftButton).toBeTruthy();
    expect(rightButton).toBeTruthy();

    leftButton.focus();
    expect(document.activeElement).toBe(leftButton);

    rightButton.focus();
    expect(document.activeElement).toBe(rightButton);
  });
});
