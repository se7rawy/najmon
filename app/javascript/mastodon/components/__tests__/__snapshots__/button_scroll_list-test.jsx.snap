// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonScrollList /> handles a large number of children correctly 1`] = `
<div
  className="button-scroll-list-container"
>
  <button
    aria-label="Scroll left"
    className="icon-button column-header__setting-btn"
    onClick={[Function]}
  >
    <div>
      MockIcon
    </div>
  </button>
  <div
    className="button-scroll-list"
  >
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
  </div>
  <button
    aria-label="Scroll right"
    className="icon-button column-header__setting-btn"
    onClick={[Function]}
  >
    <div>
      MockIcon
    </div>
  </button>
</div>
`;

exports[`<ButtonScrollList /> handles a single child correctly 1`] = `
<div
  className="button-scroll-list-container"
>
  <button
    aria-label="Scroll left"
    className="icon-button column-header__setting-btn"
    onClick={[Function]}
  >
    <div>
      MockIcon
    </div>
  </button>
  <div
    className="button-scroll-list"
  >
    <div>
      <div />
    </div>
  </div>
  <button
    aria-label="Scroll right"
    className="icon-button column-header__setting-btn"
    onClick={[Function]}
  >
    <div>
      MockIcon
    </div>
  </button>
</div>
`;

exports[`<ButtonScrollList /> renders an empty button scroll list element 1`] = `null`;

exports[`<ButtonScrollList /> renders the children 1`] = `
<div
  className="button-scroll-list-container"
>
  <button
    aria-label="Scroll left"
    className="icon-button column-header__setting-btn"
    onClick={[Function]}
  >
    <div>
      MockIcon
    </div>
  </button>
  <div
    className="button-scroll-list"
  >
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
    <div>
      <div />
    </div>
  </div>
  <button
    aria-label="Scroll right"
    className="icon-button column-header__setting-btn"
    onClick={[Function]}
  >
    <div>
      MockIcon
    </div>
  </button>
</div>
`;
